### ISSUE TITLE
featreq: function to filter out overlay packages for supported platforms

### ISSUE THREAD

**Issue by colemickens** [numtide/flake-utils#4]
This is what I'm doing now. I know there are more sophisticated versions around that will check dependencies recursively. This seems eligible for `flake-utils`, any overlay or package set providers will likely want to use some sort of function like this.

(Bonus points if we discuss whether to filter unsupported platforms out entirely, or if we give a more helpful error in those cases.)

**Comment by colemickens** [comment]
```nix
{
      packages = let
        pkgs = sys: 
          let
            nixpkgs_ = (pkgsFor inputs.nixpkgs sys true);
            packagePlatforms = pkg: pkg.meta.hydraPlatforms or pkg.meta.platforms or [ "x86_64-linux" ];
            pred = n: v: builtins.elem sys (packagePlatforms v);
          in
            nixpkgs_.lib.filterAttrs pred nixpkgs_.waylandPkgs; 
      in {
          x86_64-linux = pkgs "x86_64-linux";
          aarch64-linux = pkgs "aarch64-linux";
        };
}
```

**Comment by zimbatm** [comment]
With today's flake-utils you would achieve something similar like this:
```nix
{
  description = "colemicken's wayland overlay";

  inputs.flake-utils.url = "github:numtide/flake-utils";

  outputs = { self, nixpkgs, flake-utils }:
    {
      overlay = import ./overlay.nix;
    }
    //
    (
      flake-utils.lib.eachSystem [ "aarch64-linux" "x86_64-linux" ] (system:
        let
          pkgs = import nixpkgs {
            inherit system;
            config = { };
            overlays = [ self.overlay ];
          };
        in
        {
          packages = pkgs.waylandPkgs;
        }
      )
    );
}
```

Assuming that the overlay puts all its packages under `pkgs.waylandPkgs`, and that `pkgs.waylandPkgs` *only* contains derivations (otherwise `nix flake check` complains).

The big chunk at the end could probably be simplified for the common case.

```nix
{
  description = "colemicken's wayland overlay";

  inputs.flake-utils.url = "github:numtide/flake-utils";

  outputs = { self, nixpkgs, flake-utils }:
    flake-utils.lib.buildPackages {
      inherit self nixpkgs;
      overlay = import ./overlay.nix;
      overlayAttr = "waylandPkgs";
      systems = [ "aarch64-linux" "x86_64-linux" ];
    };
}
```

How does that look?

**Comment by zimbatm** [comment]
https://github.com/numtide/flake-utils/pull/5

**Comment by colemickens** [comment]
Cool, I need to try it and glance at `eachSystem` but it looks like what I want, yeah.  And I like the looks of #5. I'm going to try to re-visit my various flake repos next weekend, maybe dogfood #5 and then comment again / close this out.

**Comment by zimbatm** [comment]
Looking forward to your feedback!

**Comment by colemickens** [comment]
I'm skimming through the source and #5,  I'm not seeing anything that does the filtering I'm looking for based on the `meta.platforms` unless I'm looking in the wrong place?

**Comment by zimbatm** [comment]
You're right. .flattenTree should be changed to take the system as an argument. Do you want to tackle this?

**Comment by colemickens** [comment]
I can't anytime soon. But cleaning up my flake.nixs will eventually block me and I plan to clean them via this repo, so I'll get to it eventually if you don't. Thanks for confirming my understanding and starting up this repo.

**Comment by dhess** [comment]
We're slowly converting our overlays to flakes, and something that filters the per-system package set based on `meta.hydraPlatforms` would be very helpful.

In our current overlays, we import `nixpkgs + "/pkgs/top-level/release-lib.nix"` and use its functionality to build Hydra jobsets. Nix's built-in flake support, combined with `flake-utils`, replaces most of that machinery, *except* for the bit that filters out packages that don't support a particular platform.

**Comment by zimbatm** [comment]
^ please try out this new implementation in PR 13

**Comment by blaggacao** [comment]
@nrdxp [do stars align here?](https://github.com/nrdxp/nixflk/blob/a0b9f9ffe688c86bfc04c35655d783289e3484ba/flake.nix#L70-L78) (through PR13). I think nrdxp/nixflk is a complement use case to look for. It's flake is already quite complex while trying to solve common use cases around nixos configurations.

**Comment by nrdxp** [comment]
My solution does check the buildInputs of each flake package for platform dependants, but not recursively.
